File: angular/src/app/shared/shop-print/models/shop-print-document-types.ts
// Mirrors EHub.Application.Contracts/ShopManagement/PrintTemplates/ShopPrintDocumentTypes.cs
// EXACTLY - must be kept in sync by hand with the backend const class. This is the frontend-side
// half of the allowlist: the print-preview route rejects any documentType not in this list
// BEFORE ever calling the API, as defense in depth alongside the backend's own allowlist check.
export const SHOP_PRINT_DOCUMENT_TYPES = {
  Sale: 'sale',
  SaleReturn: 'sale-return',
  PurchaseOrder: 'purchase-order',
  GoodsReceipt: 'goods-receipt',
  PurchaseReturn: 'purchase-return',
  CustomerPayment: 'customer-payment',
  SupplierPayment: 'supplier-payment',
  CustomerLedgerStatement: 'customer-ledger-statement',
  SupplierLedgerStatement: 'supplier-ledger-statement',
  ExpenseVoucher: 'expense-voucher',
  CashClosingSlip: 'cash-closing-slip',
  BankTransaction: 'bank-transaction',
  BankTransfer: 'bank-transfer',
  StockAdjustment: 'stock-adjustment',
  StockCount: 'stock-count',
  StockTransaction: 'stock-transaction',
  ProductBarcodeLabel: 'product-barcode-label',
  BatchExpiryLabel: 'batch-expiry-label',
} as const;

export type ShopPrintDocumentType = (typeof SHOP_PRINT_DOCUMENT_TYPES)[keyof typeof SHOP_PRINT_DOCUMENT_TYPES];

export const SHOP_PRINT_DOCUMENT_TYPE_LIST: string[] = Object.values(SHOP_PRINT_DOCUMENT_TYPES);

export function isValidShopPrintDocumentType(value: string | null | undefined): value is ShopPrintDocumentType {
  return !!value && SHOP_PRINT_DOCUMENT_TYPE_LIST.includes(value);
}

// Document types that render as a fixed-size label (grid of labels) rather than a receipt.
export const SHOP_PRINT_LABEL_DOCUMENT_TYPES: string[] = [
  SHOP_PRINT_DOCUMENT_TYPES.ProductBarcodeLabel,
  SHOP_PRINT_DOCUMENT_TYPES.BatchExpiryLabel,
];

File: src/EHub.Application.Contracts/ShopManagement/PrintTemplates/ShopPrintDocumentTypes.cs
namespace EHub.ShopManagement.PrintTemplates;

// Backend allowlist of printable document types. The Angular frontend keeps its own mirrored
// constant array (shop-print-document-types.ts) for client-side validation before ever calling
// the API - both lists must be kept in sync by hand since these values cannot be shared directly
// across the .NET/TypeScript boundary.
public static class ShopPrintDocumentTypes
{
    public const string Sale = "sale";
    public const string SaleReturn = "sale-return";
    public const string PurchaseOrder = "purchase-order";
    public const string GoodsReceipt = "goods-receipt";
    public const string PurchaseReturn = "purchase-return";
    public const string CustomerPayment = "customer-payment";
    public const string SupplierPayment = "supplier-payment";
    public const string CustomerLedgerStatement = "customer-ledger-statement";
    public const string SupplierLedgerStatement = "supplier-ledger-statement";
    public const string ExpenseVoucher = "expense-voucher";
    public const string CashClosingSlip = "cash-closing-slip";
    public const string BankTransaction = "bank-transaction";
    public const string BankTransfer = "bank-transfer";
    public const string StockAdjustment = "stock-adjustment";
    public const string StockCount = "stock-count";
    public const string StockTransaction = "stock-transaction";
    public const string ProductBarcodeLabel = "product-barcode-label";
    public const string BatchExpiryLabel = "batch-expiry-label";

    public static readonly string[] All =
    {
        Sale, SaleReturn, PurchaseOrder, GoodsReceipt, PurchaseReturn,
        CustomerPayment, SupplierPayment, CustomerLedgerStatement, SupplierLedgerStatement,
        ExpenseVoucher, CashClosingSlip, BankTransaction, BankTransfer,
        StockAdjustment, StockCount, StockTransaction,
        ProductBarcodeLabel, BatchExpiryLabel
    };
}
